Assets/MRTK/SDK/Experimental/Drawstring/Scripts/TwoHandleManipulator.cs

```diff
@@ -0,0 +1,937 @@
+        {
+            Vector3 sum = Vector3.zero;
+            int numControllers = 0;
+            foreach (var p in pointerIdToPointerMap.Values)
+            {
+                // Check pointer has a valid controller (e.g. gaze pointer doesn't)
+                if (p.pointer.Controller != null)
+                {
+                    numControllers++;
+                    sum += p.pointer.Controller.Velocity;
+                }
+            }
+            return sum / Math.Max(1, numControllers);
+        }
+
+        private Vector3 GetPointersAngularVelocity()
+        {
+            Vector3 sum = Vector3.zero;
+            int numControllers = 0;
+            foreach (var p in pointerIdToPointerMap.Values)
+            {
+                // Check pointer has a valid controller (e.g. gaze pointer doesn't)
+                if (p.pointer.Controller != null)
+                {
+                    numControllers++;
+                    sum += p.pointer.Controller.AngularVelocity;
+                }
+            }
+            return sum / Math.Max(1, numControllers);
+        }
+
+        private bool IsNearManipulation()
+        {
+            foreach (var item in pointerIdToPointerMap)
+            {
+                if (item.Value.IsNearPointer)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+        #endregion Private Methods
+
+        #region Public Methods
+
+        /// <summary>
+        /// Releases the object that is currently manipulated
+        /// </summary>
+        public void ForceEndManipulation()
+        {
+            // end manipulation
+            if (isManipulationStarted)
+            {
+                HandleManipulationEnded(GetPointersGrabPoint(), GetPointersVelocity(), GetPointersAngularVelocity());
+            }
+            pointerIdToPointerMap.Clear();
+        }
+
+        /// <summary>
+        /// Gets the grab point for the given pointer id.
+        /// Only use if you know that your given pointer id corresponds to a pointer that has grabbed
+        /// this component.
+        /// </summary>
+        public Vector3 GetPointerGrabPoint(uint pointerId)
+        {
+            Assert.IsTrue(pointerIdToPointerMap.ContainsKey(pointerId));
+            return pointerIdToPointerMap[pointerId].GrabPoint;
+        }
+
+        #endregion Public Methods
+
+        #region Hand Event Handlers
+
+        /// <inheritdoc />
+        public void OnPointerDown(MixedRealityPointerEventData eventData)
+        {
+            if (eventData.used ||
+                (!allowFarManipulation && eventData.Pointer as IMixedRealityNearPointer == null))
+            {
+                return;
+            }
+
+            // If we only allow one handed manipulations, check there is no hand interacting yet. 
+            if (manipulationType != ManipulationHandFlags.OneHanded || pointerIdToPointerMap.Count == 0)
```

Comment:
This definitely works, though you may want to consider having some more explicit thing like:

if (Enum.HasFlag(manipulationType, TwoHanded) ||
    Enum.HasFlag(manipulationType, OneHanded) && pointerIdToPointerMap.Count == 0)

Reply:
It aligns with the comments that are written and also handles the case when nothing is set (for whatever reason)